crates/next-api/src/next_server_nft.rs

```diff
@@ -0,0 +1,340 @@
+use turbopack_core::{
+    asset::{Asset, AssetContent},
+    context::AssetContext,
+    file_source::FileSource,
+    module::Module,
+    output::{OutputAsset, OutputAssets},
+    reference_type::{CommonJsReferenceSubType, ReferenceType},
+    resolve::{ExternalType, origin::PlainResolveOrigin, parse::Request},
+    traced_asset::TracedAsset,
+};
+use turbopack_ecmascript::resolve::cjs_resolve;
+
+use crate::{nft_json::all_assets_from_entries_filtered, project::Project};
+
+#[instrument(level = Level::INFO, skip_all)]
+#[turbo_tasks::function]
+pub async fn next_server_nft_assets(project: Vc<Project>) -> Result<Vc<OutputAssets>> {
+    let is_standalone = *project.next_config().is_standalone().await?;
+    let has_next_support = *project.next_config().ci_has_next_support().await?;
+
+    let asset_context = Vc::upcast(externals_tracing_module_context(
+        ExternalType::CommonJs,
+        get_tracing_compile_time_info(),
+    ));
+
+    let project_path = project.project_path().owned().await?;
+
+    let next_resolve_origin = Vc::upcast(PlainResolveOrigin::new(
+        asset_context,
+        get_next_package(project_path.clone()).await?.join("_")?,
+    ));
+
+    let resolve_entry = async |path: &str| {
+        Ok(cjs_resolve(
+            next_resolve_origin,
+            Request::parse_string(path.into()),
+            CommonJsReferenceSubType::Undefined,
+            None,
+            false,
+        )
+        .primary_modules()
+        .await?
+        .into_iter()
+        .map(|m| **m))
+    };
+
+    let cache_handler = project
+        .next_config()
+        .cache_handler(project_path.clone())
+        .await?;
+    let cache_handlers = project
+        .next_config()
+        .experimental_cache_handlers(project_path.clone())
+        .await?;
+
+    // These are used by packages/next/src/server/require-hook.ts
+    let shared_entries: Vec<Vc<Box<dyn Module>>> =
+        ["styled-jsx", "styled-jsx/style", "styled-jsx/style.js"]
+            .into_iter()
+            .map(resolve_entry)
+            .try_flat_join()
+            .await?;
+
+    let cache_handler_entries: Vec<Vc<Box<dyn Module>>> = cache_handler
+        .into_iter()
+        .chain(cache_handlers.into_iter())
+        .map(|f| {
+            asset_context
+                .process(
+                    Vc::upcast(FileSource::new(f.clone())),
+                    ReferenceType::CommonJs(CommonJsReferenceSubType::Undefined),
+                )
+                .module()
+        })
+        .collect();
+
+    let server_entries = shared_entries
+        .iter()
+        .chain(cache_handler_entries.iter())
+        .copied()
+        .chain(if is_standalone {
+            Either::Left(
+                resolve_entry("next/dist/server/lib/start-server")
+                    .await?
+                    .chain(resolve_entry("next/dist/server/next").await?)
+                    .chain(resolve_entry("next/dist/server/require-hook").await?),
```

Comment:
If there's a way that we could put all of these paths into a JSON file, we could share that across the JS and Rust (using `include!()`) implementations.